extra/qcom_gpu_driver/qcom.py

```diff
@@ -0,0 +1,174 @@
+    if reg: self.q += [pkt4_hdr(reg, len(vals)), *vals]
+    return self
+
+  def get(self):
+    return array.array('I', self.q)
+
+class Qcom():
+  def __init__(self):
+    self.fd = os.open('/dev/kgsl-3d0', os.O_RDWR)
+
+  def _ioctl(self, nr, arg):
+    ret = fcntl.ioctl(self.fd, (3 << 30) | (ctypes.sizeof(arg) & 0x1FFF) << 16 | 0x9 << 8 | (nr & 0xFF), arg)
+    if ret != 0: raise RuntimeError(f"ioctl returned {ret}")
+    return ret
+  
+  def alloc(self, size: int, flags: int, mapped:bool=False, id:bool=False):
+    gpuobj_alloc = msm_kgsl.struct_kgsl_gpuobj_alloc(size=size, flags=flags)
+    self._ioctl(0x45, gpuobj_alloc) # IOCTL_KGSL_GPUOBJ_ALLOC
+    if not gpuobj_alloc.mmapsize:
+      raise RuntimeError(f"failed to alloc, mmap_size={gpuobj_alloc.mmapsize}")
+    gpuobj_info = msm_kgsl.struct_kgsl_gpuobj_info(id=gpuobj_alloc.id)
+    self._ioctl(0x47, gpuobj_info) # IOCTL_KGSL_GPUOBJ_INFO
+    if not mapped:
+      return gpuobj_info.id, gpuobj_info.gpuaddr
+    return gpuobj_info.id, gpuobj_info.gpuaddr, mmap.mmap(self.fd, gpuobj_alloc.mmapsize, offset=gpuobj_info.id * 0x1000)
+
+  def submit_cmds(self, ctx_id: int, *cmds: CMDBuffer):
+    objs = (msm_kgsl.struct_kgsl_command_object * len(cmds))()
+    for i, cmd in enumerate(cmds):
+      cmdbytes = cmd.get()
+      _, addr, buff = self.alloc((len(cmdbytes) * 4 + 4096 - 1) // 4096 * 4096, 0xC0A00, mapped=True)
+      ctypes.memmove(ptr(buff), ptr(cmdbytes), len(cmdbytes) * 4)
+      objs[i].gpuaddr = addr
+      objs[i].size = len(cmdbytes) * 4
+      objs[i].flags = 0x00000001
+
+    submit_req = msm_kgsl.struct_kgsl_gpu_command()
+    submit_req.flags = 0x0 # 0x11
+    submit_req.cmdlist = ctypes.addressof(objs)
+    submit_req.cmdsize = ctypes.sizeof(msm_kgsl.struct_kgsl_command_object)
+    submit_req.numcmds = len(cmds)
+    submit_req.context_id = ctx_id
+
+    self._ioctl(0x4A, submit_req)
+
+  def test(self):
+    print('****** create eye Tensor and copy it to the GPU ******')
+    _, tensor_gpuaddr, tensor_buff = self.alloc(0x144000, 0xC0C0E00, mapped=True)
+    matrix_bytes = bytearray(numpy.eye(576, dtype=numpy.float32).flatten().tobytes())
+    ctypes.memmove(ptr(tensor_buff), ptr(matrix_bytes), len(matrix_bytes))
+
+    print('****** initial values ******')
+    print(f"values = {struct.unpack('f' * 9, tensor_buff[0:9*4])}")
+
+    print('****** alloc and fill shader ******')
+    _, shader_gpuaddr, shader_buff = self.alloc(0x1000, 0x10C0A00, mapped=True)
+    ctypes.memmove(ptr(shader_buff), ptr(bytearray(SHADER)), len(SHADER))
+
+    print('****** alloc and fill consts ******')
+    _, consts_gpuaddr, consts_buff = self.alloc(0x100000, 0xC0A00, mapped=True)
+    CONSTANTS = patch_constants(0x140, tensor_gpuaddr)
+    ctypes.memmove(ptr(consts_buff), ptr(bytearray(CONSTANTS)), len(CONSTANTS))
+
+    print('****** fill cmd buff 1 ******')
+    # Setting up GPU for the compute shader mode...
+    buff1 = CMDBuffer()
+    buff1.push(opcode=0x26) # CP_WAIT_FOR_IDLE
```

Comment:
Ideally there's ways to autogen a header file with all these constants instead of comments like this.